packages/relay-compiler/transforms/SkipRedundantNodesTransform.js

```diff
@@ -207,13 +207,19 @@ function transformNode<T: Node>(
 function sortSelections(
   selections: $ReadOnlyArray<Selection>,
 ): $ReadOnlyArray<Selection> {
-  return [...selections].sort((a, b) => {
-    return a.kind === 'InlineFragment' || a.kind === 'Condition'
-      ? 1
-      : b.kind === 'InlineFragment' || b.kind === 'Condition'
-        ? -1
-        : 0;
-  });
+  const [scalarsAndLinkedFields, rest] = selections.reduce(
+    (acc, next) => {
+      const [scalarsAndLinkedFields, rest] = acc;
+      const isScalarOrLinkedField =
+        next.kind === 'ScalarField' || next.kind === 'LinkedField';
+      if (isScalarOrLinkedField) {
+        return [[...scalarsAndLinkedFields, next], rest];
+      }
+      return [scalarsAndLinkedFields, [...rest, next]];
+    },
+    [[], []],
+  );
+  return [...scalarsAndLinkedFields, ...rest];
```

Comment:
Repeatedly creating new arrays like this is incredibly inefficient. Instead, let’s create just two new arrays and partition the elements into them. There’s a helper function partitionArray() (or similar) in RelayParser that we can extract into a module and reuse it for this.

Reply:
Done in 516ae8c